FiftyOne.Pipeline.Core/FlowElements/Pipeline.cs

```diff
@@ -197,10 +204,29 @@ public IReadOnlyList<IFlowElement> FlowElements
         {
             get
             {
-                return _elementAvailableProperties;
+                if (_elementAvailableProperties is object)
+                {
+                    return _elementAvailableProperties;
+                }
+                lock (_elementAvailablePropertiesLock)
+                {
+                    if (_elementAvailableProperties is object)
+                    {
+                        return _elementAvailableProperties;
+                    }
+                    bool hadFailures = false;
+                    var properties = GetElementAvailableProperties(_flowElements, out hadFailures);
+                    if (!hadFailures)
+                    {
+                        _elementAvailableProperties = properties;
+                    }
+                    return properties;

```

Comment:
nice implementation with partial properties.  maybe worth adding some unit tests to cover this behavior?

Reply:
Added: https://github.com/51Degrees/pipeline-dotnet/pull/46/commits/0832ff99e1891ffc9dd650e492f98a08c5d80bb0